### PR TITLE
Allow for serializing/deserialzing SparseArray2d

### PR DESCRIPTION
**Pull request by Dekken**
Referencing #137

### CODE REVIEW

File: lib/include/tick/array/sparsearray2d.h [hunk 1]
```diff
@@ -89,6 +102,43 @@ class SparseArray2d : public BaseArray2d<T> {
...
+      return spd2d;
+    }
+
+    static void TO_FILE(std::string _file, const SparseArray2d<T>& array){
+      std::ofstream ss(_file, std::ios::out | std::ios::binary);
+      {
+        cereal::PortableBinaryOutputArchive oarchive(ss);
+        oarchive(array);
+      }
+      /** Code to compare serialized versions to original
```
Comment (Mbompr) [inline lib/include/tick/array/sparsearray2d.h]: I Think this should go in a unit test rather left commented here.

File: lib/include/tick/array/sparsearray2d.h [hunk 2]
```diff
@@ -135,6 +185,62 @@ Array2d<T> BaseArray2d<T>::as_array2d() {
     return c;
 }
 
+// Using save/load methods directory on SparseArray2d<T> does not compile simply
```
Comment (Mbompr) [inline lib/include/tick/array/sparsearray2d.h]: What does this mean?

File: lib/include/tick/array/sparsearray2d.h [hunk 3]
```diff
@@ -135,6 +185,62 @@ Array2d<T> BaseArray2d<T>::as_array2d() {
...
+      s._data        = s_data;
+      s._indices     = s_indices;
+      s._row_indices = s_row_indices;
+
+      s.is_data_allocation_owned = 1;
+      s.is_indices_allocation_owned = 1;
+      s.is_row_indices_allocation_owned = 1;
+
+    }catch(const std::exception& e){
+      std::cerr << e.what() << std::endl;
```
Comment (Mbompr) [inline lib/include/tick/array/sparsearray2d.h]: maybe should we throw an exception? For example with TICK_ERROR macro

File: lib/include/tick/array/ssparsearray2d.h [hunk 4]
```diff
@@ -277,7 +277,8 @@ std::shared_ptr<SSparseArray2d<T>> SSparseArray2d<T>::new_ptr(ulong n_rows, ulon
 // Constructor from a SparseArray2d (copy is made)
 template<typename T>
 std::shared_ptr<SSparseArray2d<T>> SSparseArray2d<T>::new_ptr(SparseArray2d<T> &a) {
-    std::shared_ptr<SSparseArray2d<T>> aptr = SSparseArray2d<T>::new_ptr(a.n_rows(), a.n_rows(), a.size_sparse());
+    std::shared_ptr<SSparseArray2d<T>> aptr = SSparseArray2d<T>::new_ptr(a.n_rows(), a.n_cols(), a.size_sparse());
+    aptr->_size = a.size();
```
Comment (Mbompr) [inline lib/include/tick/array/ssparsearray2d.h]: Nice catch!

File: tick/array/tests/serializer.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+from tick.array.build.array import Serializer
+from tick.optim.solver.tests.solver import TestSolver
+
+from tick.optim.model import ModelLinReg, ModelLogReg, ModelPoisReg, ModelCoxRegPartialLik
+from tick.simulation import SimuLinReg, SimuLogReg, SimuPoisReg, SimuCoxReg, weights_sparse_gauss
+
+from scipy.sparse import csr_matrix
+
+model_simu_map = {
```
Comment (Mbompr) [inline tick/array/tests/serializer.py]: it would be cleaner if these instantiations were in the setup method of test?

File: tools/python/serializer.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# License: BSD 3 clause
```
Comment (Mbompr) [inline tools/python/serializer.py]: I would rather put this kind of utility closer to the fetch_tick_dataset method. I also think that it should be a method rather than a script.

File: tools/travis/docker_run.sh [hunk 7]
```diff
@@ -11,6 +11,12 @@ eval "$(pyenv init -)"
 pyenv global ${PYVER}
 pyenv local ${PYVER}
 
+gcc -v
+
+git log -10
+
+swig -version
```
Comment (Mbompr) [inline tools/travis/docker_run.sh]: To be removed ?

File: tools/travis/osx_run.sh [hunk 8]
```diff
@@ -2,12 +2,22 @@
...
 set -e -x
 
-shell_session_update() { :; }
+shell_session_update() { 
+  set +e
+  echo "This is so travis doesn't break"
+  return 0;
+}
+
+rvm get head
```
Comment (Mbompr) [inline tools/travis/osx_run.sh]: I think that everything that is related to Travis compilation should be done in another PR.

File: lib/include/tick/array/serializer.h [hunk 9]
```diff
@@ -0,0 +1,20 @@
+
+
+#include "sparsearray.h"
+#include "sparsearray2d.h"
+
+class Serializer{
```
Comment (Mbompr) [inline lib/include/tick/array/serializer.h]: Do we really still need this class? As it has no state we should be able to expose simply array_to_file and sparse2d_to_file methods.

File: lib/include/tick/array/sparsearray2d.h (hunk 1 above)
Comment (Mbompr) [reply lib/include/tick/array/sparsearray2d.h]: A cpp unit test

Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: It's a bit more complicated that that as it requires python to be run first but yeah can be

File: lib/include/tick/array/sparsearray2d.h (hunk 2 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: The save/load functions for SparseArray2d are global and not member functions of the class SparseArray2d, which would not compile.

File: lib/include/tick/array/sparsearray2d.h (hunk 3 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: sure

File: tools/python/serializer.py (hunk 6 above)
Comment (Dekken) [reply tools/python/serializer.py]: sure

File: tools/travis/osx_run.sh (hunk 8 above)
Comment (Dekken) [reply tools/travis/osx_run.sh]: will be, travis is slow so I was testing multiple things at once

File: tools/travis/docker_run.sh (hunk 7 above)
Comment (Dekken) [reply tools/travis/docker_run.sh]: sure, I think it's useful for debugging tho

File: lib/include/tick/array/sparsearray2d.h (hunk 1 above)
Comment (Mbompr) [reply lib/include/tick/array/sparsearray2d.h]: We can probably make this run in C++ directly by using this sample matrix
```
   _data  = [ 5 8 3 6 ]
   _row_indices = [ 0 0 2 3 4 ]
   _indices = [ 0 1 2 1 ]
```
(from https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_.28CSR.2C_CRS_or_Yale_format.29)

File: lib/include/tick/array/sparsearray2d.h (hunk 2 above)
Comment (Mbompr) [reply lib/include/tick/array/sparsearray2d.h]: Hum, I still don't get it sorry :). They won't compile because they are global and not member, or they need to be global and not member because otherwise they won't compile?
There is probably something I ignore about this compilation problem

File: lib/include/tick/array/serializer.h (hunk 9 above)
Comment (Dekken) [reply lib/include/tick/array/serializer.h]: I'm not a fan of global public functions but I guess for swig it makes sense.

File: lib/include/tick/array/sparsearray2d.h (hunk 2 above)
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: They need to be global, which is consistent with array.h save/load methods.

Comment (Mbompr) [reply lib/include/tick/array/sparsearray2d.h]: Oh ok, then maybe can we cahnge comment to
> save/load methods need to be global to be consistent with array.h save/load methods.

File: tools/travis/docker_run.sh (hunk 7 above)
Comment (Mbompr) [reply tools/travis/docker_run.sh]: Ok, as you told me we better keep those :)

File: tools/travis/osx_run.sh (hunk 8 above)
Comment (Mbompr) [reply tools/travis/osx_run.sh]: Ok!

File: lib/src/cpp-test/array/serializer.cpp [hunk 10]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (Mbompr) [inline lib/src/cpp-test/array/serializer.cpp]: Can we make a gtest of it?

File: lib/src/swig/array/serializer.i [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+%include <std_string.i>
+%include <std_shared_ptr.i>
+
+%{
+#include "tick/array/serializer.h"
+%}
+
+// We need to expose this empty class to Python, so that we can pass
+// shared pointers of it to models
```
Comment (Mbompr) [inline lib/src/swig/array/serializer.i]: Deprecated comment :)

File: setup.py [hunk 12]
```diff
@@ -283,7 +283,7 @@ def add_dir_name(dir_name, filenames):
                               '-Ilib/include',
```
Comment (Mbompr) [inline setup.py]: Could we avoid to modify setup.py in this PR which is already rich enough

File: sh/configure_env.sh [hunk 13]
```diff
@@ -19,7 +19,7 @@
 
 set -e
 
-TICK_CONFIGURED=1
+echo "Entering configure_env.sh"
```
Comment (Mbompr) [inline sh/configure_env.sh]: Should we leave this?

File: tick/array/tests/serializer.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    tick_double_sparse2d_to_file(features, X)
+    
+    y1 = tick_double_array_from_file(lables)
+    X1 = tick_double_sparse2d_from_file(features)
+    
+    print("X1 type: " + str(type(X1)))
+    print(X1.toarray())
+
+if __name__ == '__main__':
+  unittest.main()
```
Comment (Mbompr) [inline tick/array/tests/serializer.py]: You can use `np.test.assert_array_equal` to check array equality

File: tick/array_test/tests/array_memory_test.py [hunk 15]
```diff
@@ -2,12 +2,29 @@
...
+
+def dump_garbage():
+
+    gc.collect()
+    print("\nGARBAGE:")
+    print("\nGARBAGE OBJECTS:")
+    for x in gc.garbage:
+        s = str(x)
+        if len(s) > 80: s = s[:80]
+        print(type(x),"\n  ", s)
```
Comment (Mbompr) [inline tick/array_test/tests/array_memory_test.py]: Tests should not print :)

File: tick/dataset/serializer.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
+# License: BSD 3 clause
```
Comment (Mbompr) [inline tick/dataset/serializer.py]: This could be added as a utility to the dataset module

File: lib/src/cpp-test/array/serializer.cpp (hunk 10 above)
Comment (Dekken) [reply lib/src/cpp-test/array/serializer.cpp]: Not strictly, it would require a pre executed python step to create the cereal binary file.

File: lib/include/tick/array/sparsearray2d.h [hunk 17]
```diff
@@ -89,6 +102,33 @@ class SparseArray2d : public BaseArray2d<T> {
...
+
+    static SparseArray2d<T> FROM_FILE(const std::string& _file){
+      SparseArray2d<T> spd2d;
+      {
+        std::ifstream data(_file, std::ios::in | std::ios::binary);
+        cereal::PortableBinaryInputArchive iarchive(data);
+        iarchive(spd2d);
+      }
+      return spd2d;
+    }
```
[the remark this answers is not in the document]
Comment (Dekken) [reply lib/include/tick/array/sparsearray2d.h]: I've been using it from C++, but you're right that it's not needed for python

File: sh/configure_env.sh (hunk 13 above)
Comment (Dekken) [reply sh/configure_env.sh]: It was moved to the end of the file

File: tick/array/tests/serializer.py (hunk 14 above)
Comment (Dekken) [reply tick/array/tests/serializer.py]: That function is never called, it was for debugging GC while trying to figure out if SparseArray2d was being deleted properly. Can remove though,

File: tick/array_test/tests/array_memory_test.py (hunk 15 above)
Comment (Dekken) [reply tick/array_test/tests/array_memory_test.py]: That function is never called, it was for debugging GC while trying to figure out if SparseArray2d was being deleted properly. Can remove though,

File: tick/dataset/serializer.py (hunk 16 above)
Comment (Dekken) [reply tick/dataset/serializer.py]: Is it not?

File: setup.py (hunk 12 above)
Comment (Dekken) [reply setup.py]: Sure, do you plan on fixing cpplint as part of flattening? :P

File: tick/array/tests/serializer.py (hunk 14 above)
Comment (Dekken) [reply tick/array/tests/serializer.py]: k
